L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,463 @@
+class FileWatcherCertificateProvider final
+    : public CertificateProviderInterface {
+ public:
+  // Constructor to get credential updates from root and identity file paths.
+  //
+  // @param private_key_path is the file path of the private key.
+  // @param identity_certificate_path is the file path of the identity
+  // certificate chain.
+  // @param root_cert_path is the file path to the root certificate bundle.
+  // @param refresh_interval_sec is the refreshing interval that we will check
+  // the files for updates.
+  FileWatcherCertificateProvider(const std::string& private_key_path,
+                                 const std::string& identity_certificate_path,
+                                 const std::string& root_cert_path,
+                                 unsigned int refresh_interval_sec);
+
+  // Constructor to get credential updates from identity file paths only.
+  FileWatcherCertificateProvider(const std::string& private_key_path,
+                                 const std::string& identity_certificate_path,
+                                 unsigned int refresh_interval_sec);
+
+  // Constructor to get credential updates from root file path only.
+  FileWatcherCertificateProvider(const std::string& root_cert_path,
+                                 unsigned int refresh_interval_sec);
+    }
+
+
+```
+### Custom Verification
+We aim to provide distinct interfaces for chain building customization and for additional validation of an already built chain. These two features have distinctly different expected users. Doing some extra validation based on the contents of the peer certificate chain is an operation that anyone doing (m)TLS might be interested in, and it should not and does not require any deep X.509 expertise. Replacing how chain building works is a relatively rare requirement and it does require deep X.509 expertise to do correctly. We want extra validation to be easy and clear for users to add without worrying about interacting with chain building itself. On the other hand, chain building is complex to implement, and we suspect that most users will stay with the default behavior. However, to enable advanced use, such as dynamic selection of the trust bundle based on the peer certificate (e.g. SPIFFE federation), this is needed. 
+#### Post Handshake Verification
+```c++
+// Contains the verification-related information associated with a connection
+// request. Users should not directly create or destroy this request object, but
+// shall interact with it through CertificateVerifier's Verify() and Cancel().
+// TODO(gtcooke94) Add expected formats for all string values here.
+class TlsCustomVerificationCheckRequest {
+ public:
+  explicit TlsCustomVerificationCheckRequest(
+      grpc_tls_custom_verification_check_request* request);
+  ~TlsCustomVerificationCheckRequest() {}
+
+  absl::string_view target_name() const;
+  absl::string_view peer_cert() const;
```

Comment:
I'll add a TODO for myself to look into this too - something in my memory from months ago wants to say that, in the case that we are sourcing this value from an OpenSSL API that _might_ return a verified chain. I need to double check on this. But it might just be the unvalidated presented chain.

Either way, since we source the root cert subject after it's been verified, we can certainly explicitly provide the verified chain as well.

Reply:
But I do wonder, if this is the case, should we remove this field?

Reply:
+1 I think we should either remove this field or replace its implementation so that it returns the verified chain instead.

Reply:
Friendly ping

Reply:
I dug in a little more, this comment explains why it is not guaranteed:
https://github.com/grpc/grpc/blob/c449b796126554cc4d6a8b0b6957ae4077d33c8b/include/grpc/grpc_security_constants.h#L34-L45

Given that, WDYT?

Reply:
I think that comment is another signal that this API is not useful for anyone since they cannot really predict what the return value should be. I think we should probably deprecate + delete this API and replace it with APIs that return predictable and explicit values.

Reply:
I think that, given the structure of this, we could be certain that we get a verified chain. Otherwise, we won't make it to the point that we'll call this custom verifier.

I think in this case we could probably keep this as `peer_cert_full_chain` and change the comments on what we fill it with? Because it's already on the `VerifiedPeerCertificateChainInfo` object, so it would make sense that `VerifiedPeerCertificateChainInfo.peer_cert_full_chain` would be verfied.

WDYT?

Reply:
How do we get a verified chain on resumed handshakes?

Reply:
Can we? IIRC resumed handshakes skip verification?

Reply:
Ya that's what I was hinting at. Resumed handshakes skip the cryptographic verification, but I'm not sure if they skip the post-handshake verification. If that's correct, we need to explain that some of these fields may not be available on resumed handshakes.

Reply:
Coming at this from another angle - regardless of the current behavior, do we want it to do custom verification on resumption or say resumption is resumption and _no_ verification will happen?

Reply:
Great point. I think skipping verification on resumption may be the way to go, since that's already the existing behavior for chain building.

Reply:
So the end state of this is as follows:

This field can be called `verified_blahblahblah` because this gets used as part of the custom verification callback. We will implement such that those won't be triggered on resumption because that's not really a sensible thing to do.

Now what remains is - I believe that we design to failout before the post-handshake custom verification if the regular verification fails, it's at a totally different layer than the openssl verification that is happening.
So I think that if we get to here we should be guaranteed that we have a verified chain? Does that all check out to you? Let me know and I'll update this section of the comments/docs

Reply:
Two follow-up questions:
1. To confirm, the `CertificateVerifierInterface::Verify` will not be executed at all on resumed handshakes - is that correct? If yes, can we add a new comment paragraph on this method that clarifies this?
2. On what to do if regular verification fails: if regular verification fails, the TLS handshake fails, so we should not even get to the point where we can call `CertificateVerifierInterface::Verify`.

Reply:
Friendly ping.

Reply:
Added explanation on the `Verify` method itself

1. Correct
2. Correct